internal/interfaces/grpc/handler/operator.go

```diff
@@ -987,511 +787,180 @@ func (o operatorHandler) marketFragmenterSplitFunds(
+) (*daemonv2.RemoveWebhookResponse, error) {
+	if err := h.operatorSvc.RemoveWebhook(ctx, req.GetId()); err != nil {
+		return nil, err
 	}
+	return &daemonv2.RemoveWebhookResponse{}, nil
 }
 
-func parseBalance(bal *tdexv1.Balance) application.Balance {
-	var baseAmount, quoteAmount uint64
-	if bal != nil {
-		baseAmount = bal.GetBaseAmount()
-		quoteAmount = bal.GetQuoteAmount()
+func (h *operatorHandler) listWebhooks(
+	ctx context.Context, req *daemonv2.ListWebhooksRequest,
+) (*daemonv2.ListWebhooksResponse, error) {
+	actionType, err := parseWebhookActionType(req.GetAction())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-	return application.Balance{baseAmount, quoteAmount}
-}
 
-func parseFixedFee(fee *tdexv1.Fixed) application.Fee {
-	var baseFee, quoteFee int64
-	if fee != nil {
-		baseFee = fee.GetBaseFee()
-		quoteFee = fee.GetQuoteFee()
-	}
-	return application.Fee{
-		FixedBaseFee:  baseFee,
-		FixedQuoteFee: quoteFee,
+	hooks, err := h.operatorSvc.ListWebhooks(ctx, actionType)
+	if err != nil {
+		return nil, err
 	}
+	return &daemonv2.ListWebhooksResponse{
+		WebhookInfo: hooksInfo(hooks).toProto(),
+	}, nil
 }
 
-func parsePrice(p *tdexv1.Price) (price application.Price, err error) {
-	var basePrice, quotePrice = decimal.NewFromInt(0), decimal.NewFromInt(0)
-	if p != nil {
-		basePrice = decimal.NewFromFloat(p.GetBasePrice())
-		quotePrice = decimal.NewFromFloat(p.GetQuotePrice())
-	}
-	pp := application.Price{
-		BasePrice:  basePrice,
-		QuotePrice: quotePrice,
+func (h *operatorHandler) listDeposits(
+	ctx context.Context, req *daemonv2.ListDepositsRequest,
+) (*daemonv2.ListDepositsResponse, error) {
+	accountName, err := parseAccountName(req.GetAccountName())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-	if err = pp.Validate(); err != nil {
-		return
+	page, err := parsePage(req.GetPage())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-	price = pp
-	return
-}
 
-func parseStrategy(sType daemonv1.StrategyType) (domain.StrategyType, error) {
-	strategyType := domain.StrategyType(sType)
-	if strategyType < domain.StrategyTypePluggable ||
-		strategyType > domain.StrategyTypeUnbalanced {
-		return -1, errors.New("strategy type is unknown")
+	deposits, err := h.operatorSvc.ListDeposits(
+		ctx, accountName, page,
+	)
+	if err != nil {
+		return nil, err
 	}
-	return strategyType, nil
-}
 
-func toUtxoInfoList(list []application.UtxoInfo) []*daemonv1.UtxoInfo {
-	res := make([]*daemonv1.UtxoInfo, 0, len(list))
-	for _, u := range list {
-		res = append(res, &daemonv1.UtxoInfo{
-			Outpoint: &daemonv1.Outpoint{
-				Hash:  u.Outpoint.Hash,
-				Index: int32(u.Outpoint.Index),
-			},
-			Value: u.Value,
-			Asset: u.Asset,
-		})
-	}
-	return res
+	return &daemonv2.ListDepositsResponse{
+		AccountName: accountName,
+		Deposits:    depositsInfo(deposits).toProto(),
```

Comment:
The fact is that in core/ports we have only definitions of interface types and this makes them perfect to be used as cross-layer types, instead of having repetition of types and validation functions across layers.

> So ideally each layer should have types.go file where each type would be defined validated and where there would be conversion to layer bellow.

But I agree with you on this, and I just noticed that there are no validation checks in the app services at the moment. I'm going to add them. I have this idea of adding the validation functions in core/ports so every layer can just make use of them instead of implementing their own methods.

What do you think of this approach? it seems it solves both the problems you stated and the one i wanted to solve which was preventing to have basically lot of duplicated code in all layers.

Reply:
I think the cleanest(and yeah most boring way) would be to have types for each layer.

Reply:
What I don't really like about this pattern is that we end up having lots of code repetition, especially for validation functions, and this can easily introduce bugs and, in general, makes the codebase harder to maintain too.

On the other side, expecting to have only interface types in ports make them usable as cross-layer types. We end up having validation functions only on higher layers (interfaces), where we want to validate what the user is passing in the request message, while on lower ones we use only internal concrete types, where we are sure to use only correct values.